# Processing Tem


In [1]:
import numpy as np
import pandas as pd
from pandas.tseries.offsets import DateOffset
from pandas import Timestamp

### Import data

In [2]:
weather_data = pd.read_csv("./data/weatherAUS.csv")

### Exploring Data

In [3]:
weather_data.describe()

MinTemp        MaxTemp       Rainfall   Evaporation  \
count  141556.000000  141871.000000  140787.000000  81350.000000   
mean       12.186400      23.226784       2.349974      5.469824   
std         6.403283       7.117618       8.465173      4.188537   
min        -8.500000      -4.800000       0.000000      0.000000   
25%         7.600000      17.900000       0.000000      2.600000   
50%        12.000000      22.600000       0.000000      4.800000   
75%        16.800000      28.200000       0.800000      7.400000   
max        33.900000      48.100000     371.000000    145.000000   

           Sunshine  WindGustSpeed   WindSpeed9am   WindSpeed3pm  \
count  74377.000000  132923.000000  140845.000000  139563.000000   
mean       7.624853      39.984292      14.001988      18.637576   
std        3.781525      13.588801       8.893337       8.803345   
min        0.000000       6.000000       0.000000       0.000000   
25%        4.900000      31.000000       7.000000      13.000000   
50%        8.500000      39.000000      13.000000      19.000000   
75%       10.600000      48.000000      19.000000      24.000000   
max       14.500000     135.000000     130.000000      87.000000   

         Humidity9am    Humidity3pm    Pressure9am    Pressure3pm  \
count  140419.000000  138583.000000  128179.000000  128212.000000   
mean       68.843810      51.482606    1017.653758    1015.258204   
std        19.051293      20.797772       7.105476       7.036677   
min         0.000000       0.000000     980.500000     977.100000   
25%        57.000000      37.000000    1012.900000    1010.400000   
50%        70.000000      52.000000    1017.600000    1015.200000   
75%        83.000000      66.000000    1022.400000    1020.000000   
max       100.000000     100.000000    1041.000000    1039.600000   

           Cloud9am      Cloud3pm        Temp9am        Temp3pm        RISK_MM  
count  88536.000000  85099.000000  141289.000000  139467.000000  142193.000000  
mean       4.437189      4.503167      16.987509      21.687235       2.360682  
std        2.887016      2.720633       6.492838       6.937594       8.477969  
min        0.000000      0.000000      -7.200000      -5.400000       0.000000  
25%        1.000000      2.000000      12.300000      16.600000       0.000000  
50%        5.000000      5.000000      16.700000      21.100000       0.000000  
75%        7.000000      7.000000      21.600000      26.400000       0.800000  
max        9.000000      9.000000      40.200000      46.700000     371.000000

In [4]:
weather_data.head(5)

Date Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
0  2008-12-01   Albury     13.4     22.9       0.6          NaN       NaN   
1  2008-12-02   Albury      7.4     25.1       0.0          NaN       NaN   
2  2008-12-03   Albury     12.9     25.7       0.0          NaN       NaN   
3  2008-12-04   Albury      9.2     28.0       0.0          NaN       NaN   
4  2008-12-05   Albury     17.5     32.3       1.0          NaN       NaN   

  WindGustDir  WindGustSpeed WindDir9am  ... Humidity3pm  Pressure9am  \
0           W           44.0          W  ...        22.0       1007.7   
1         WNW           44.0        NNW  ...        25.0       1010.6   
2         WSW           46.0          W  ...        30.0       1007.6   
3          NE           24.0         SE  ...        16.0       1017.6   
4           W           41.0        ENE  ...        33.0       1010.8   

   Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  RISK_MM  \
0       1007.1       8.0       NaN     16.9     21.8         No      0.0   
1       1007.8       NaN       NaN     17.2     24.3         No      0.0   
2       1008.7       NaN       2.0     21.0     23.2         No      0.0   
3       1012.8       NaN       NaN     18.1     26.5         No      1.0   
4       1006.0       7.0       8.0     17.8     29.7         No      0.2   

   RainTomorrow  
0            No  
1            No  
2            No  
3            No  
4            No  

[5 rows x 24 columns]

In [5]:
#convert the datatype of 'Date' to Timestap type
weather_data['Date']=pd.to_datetime(weather_data['Date'])

In [6]:
weather_data.sort_values(by=['Date']).head(5)

Date  Location  MinTemp  MaxTemp  Rainfall  Evaporation  Sunshine  \
44351 2007-11-01  Canberra      8.0     24.3       0.0          3.4       6.3   
44352 2007-11-02  Canberra     14.0     26.9       3.6          4.4       9.7   
44353 2007-11-03  Canberra     13.7     23.4       3.6          5.8       3.3   
44354 2007-11-04  Canberra     13.3     15.5      39.8          7.2       9.1   
44355 2007-11-05  Canberra      7.6     16.1       2.8          5.6      10.6   

      WindGustDir  WindGustSpeed WindDir9am  ... Humidity3pm  Pressure9am  \
44351          NW           30.0         SW  ...        29.0       1019.7   
44352         ENE           39.0          E  ...        36.0       1012.4   
44353          NW           85.0          N  ...        69.0       1009.5   
44354          NW           54.0        WNW  ...        56.0       1005.5   
44355         SSE           50.0        SSE  ...        49.0       1018.3   

       Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  RainToday  RISK_MM  \
44351       1015.0       7.0       7.0     14.4     23.6         No      3.6   
44352       1008.4       5.0       3.0     17.5     25.7        Yes      3.6   
44353       1007.2       8.0       7.0     15.4     20.2        Yes     39.8   
44354       1007.0       2.0       7.0     13.5     14.1        Yes      2.8   
44355       1018.5       7.0       7.0     11.1     15.4        Yes      0.0   

       RainTomorrow  
44351           Yes  
44352           Yes  
44353           Yes  
44354           Yes  
44355            No  

[5 rows x 24 columns]

The start date is 2007

In [7]:
weather_data.sort_values(by=['Date']).tail(5)

Date       Location  MinTemp  MaxTemp  Rainfall  Evaporation  \
38814 2017-06-25     WaggaWagga     -0.1     12.7       0.0          NaN   
32833 2017-06-25         Sydney      7.6     19.3       0.0          3.4   
29496 2017-06-25       Richmond      2.7     20.0       0.2          NaN   
26545 2017-06-25        Penrith      3.1     20.0       0.0          NaN   
35838 2017-06-25  SydneyAirport      6.8     19.5       0.0          3.4   

       Sunshine WindGustDir  WindGustSpeed WindDir9am  ... Humidity3pm  \
38814       NaN         SSW           19.0        SSW  ...        54.0   
32833       9.4           W           35.0          W  ...        32.0   
29496       NaN           W           37.0        NaN  ...        35.0   
26545       NaN          SW           28.0         SE  ...        45.0   
35838       9.4         WSW           46.0         NW  ...        42.0   

       Pressure9am  Pressure3pm  Cloud9am  Cloud3pm  Temp9am  Temp3pm  \
38814       1020.8       1018.5       8.0       8.0      4.0     10.7   
32833       1018.6       1015.4       1.0       1.0      9.4     18.8   
29496       1018.8       1015.2       NaN       NaN      9.6     19.4   
26545          NaN          NaN       NaN       NaN      7.8     17.8   
35838       1018.4       1015.2       1.0       1.0     11.9     17.9   

       RainToday  RISK_MM  RainTomorrow  
38814         No      0.0            No  
32833         No      0.0            No  
29496         No      0.0            No  
26545         No      0.0            No  
35838         No      0.0            No  

[5 rows x 24 columns]

The end date is 2017

### Number of local areas in the data

In [8]:
location=weather_data['Location'].unique()
location

array(['Albury', 'BadgerysCreek', 'Cobar', 'CoffsHarbour', 'Moree',
       'Newcastle', 'NorahHead', 'NorfolkIsland', 'Penrith', 'Richmond',
       'Sydney', 'SydneyAirport', 'WaggaWagga', 'Williamtown',
       'Wollongong', 'Canberra', 'Tuggeranong', 'MountGinini', 'Ballarat',
       'Bendigo', 'Sale', 'MelbourneAirport', 'Melbourne', 'Mildura',
       'Nhil', 'Portland', 'Watsonia', 'Dartmoor', 'Brisbane', 'Cairns',
       'GoldCoast', 'Townsville', 'Adelaide', 'MountGambier', 'Nuriootpa',
       'Woomera', 'Albany', 'Witchcliffe', 'PearceRAAF', 'PerthAirport',
       'Perth', 'SalmonGums', 'Walpole', 'Hobart', 'Launceston',
       'AliceSprings', 'Darwin', 'Katherine', 'Uluru'], dtype=object)

In [9]:
information_matrix= weather_data.describe()

In [10]:
print("Number of rows in the dataset",len(weather_data))

Number of rows in the dataset 142193


The number of rows of humidity data at 9am that are not NaN is 140419  
The number of rows of humidity data at 3pm that are not NaN is 138583

In [11]:
print("Number of humidity 9am data rows that are NaN is ",len(weather_data)-information_matrix.loc['count'].loc['Humidity9am'])
print("Number of humidity 3pm data rows that are NaN is ",len(weather_data)-information_matrix.loc['count'].loc['Humidity3pm'])

Number of humidity 9am data rows that are NaN is  1774.0
Number of humidity 3pm data rows that are NaN is  3610.0


In [12]:
date_time=pd.to_datetime(weather_data['Date'])
date = date_time[0]

Because each location has different date and time range

In [13]:
start_date=weather_data.groupby('Location',as_index=False)['Date'].min()
start_date['Date']=pd.to_datetime(start_date['Date'])

In [14]:
end_date=weather_data.groupby('Location',as_index=False)['Date'].max()
end_date['Date']=pd.to_datetime(end_date['Date'])

## Filling the missing humidity9am data

Since there are some missing date in our data so we need to fill the missing dates and other values will be NaN

In [15]:
temp_weather_data = pd.DataFrame()
for each_location in location:
    weather_location_data = weather_data.loc[weather_data['Location']==each_location]
    range_of_date=idx=pd.date_range(pd.to_datetime(start_date.loc[start_date['Location']==each_location,'Date'].values[0]),pd.to_datetime(end_date.loc[end_date['Location']==each_location,'Date'].values[0]))
    weather_location_data =weather_location_data.set_index('Date').reindex(idx).reset_index().rename(columns={'index':'Date'})
    weather_location_data.Location=weather_location_data.Location.fillna(each_location)
    
    temp_weather_data=pd.concat([temp_weather_data,weather_location_data])

In [16]:
print("After filling the missing date we have total rows of",len(temp_weather_data))

After filling the missing date we have total rows of 149498


In [17]:
temp_weather_data=temp_weather_data.reset_index().drop('index',1)

In [18]:
temp_weather_data.describe()

MinTemp        MaxTemp       Rainfall   Evaporation  \
count  141556.000000  141871.000000  140787.000000  81350.000000   
mean       12.186400      23.226784       2.349974      5.469824   
std         6.403283       7.117618       8.465173      4.188537   
min        -8.500000      -4.800000       0.000000      0.000000   
25%         7.600000      17.900000       0.000000      2.600000   
50%        12.000000      22.600000       0.000000      4.800000   
75%        16.800000      28.200000       0.800000      7.400000   
max        33.900000      48.100000     371.000000    145.000000   

           Sunshine  WindGustSpeed   WindSpeed9am   WindSpeed3pm  \
count  74377.000000  132923.000000  140845.000000  139563.000000   
mean       7.624853      39.984292      14.001988      18.637576   
std        3.781525      13.588801       8.893337       8.803345   
min        0.000000       6.000000       0.000000       0.000000   
25%        4.900000      31.000000       7.000000      13.000000   
50%        8.500000      39.000000      13.000000      19.000000   
75%       10.600000      48.000000      19.000000      24.000000   
max       14.500000     135.000000     130.000000      87.000000   

         Humidity9am    Humidity3pm    Pressure9am    Pressure3pm  \
count  140419.000000  138583.000000  128179.000000  128212.000000   
mean       68.843810      51.482606    1017.653758    1015.258204   
std        19.051293      20.797772       7.105476       7.036677   
min         0.000000       0.000000     980.500000     977.100000   
25%        57.000000      37.000000    1012.900000    1010.400000   
50%        70.000000      52.000000    1017.600000    1015.200000   
75%        83.000000      66.000000    1022.400000    1020.000000   
max       100.000000     100.000000    1041.000000    1039.600000   

           Cloud9am      Cloud3pm        Temp9am        Temp3pm        RISK_MM  
count  88536.000000  85099.000000  141289.000000  139467.000000  142193.000000  
mean       4.437189      4.503167      16.987509      21.687235       2.360682  
std        2.887016      2.720633       6.492838       6.937594       8.477969  
min        0.000000      0.000000      -7.200000      -5.400000       0.000000  
25%        1.000000      2.000000      12.300000      16.600000       0.000000  
50%        5.000000      5.000000      16.700000      21.100000       0.000000  
75%        7.000000      7.000000      21.600000      26.400000       0.800000  
max        9.000000      9.000000      40.200000      46.700000     371.000000

In [19]:
print("Number of humidity 9am data rows that are NaN is ",len(temp_weather_data)-temp_weather_data.describe().loc['count'].loc['Humidity9am'])

Number of humidity 9am data rows that are NaN is  9079.0


### Take a look at the missing humidity value

First find all the Nan values in Humidity9am col

In [20]:
missing_humidity9am_data = temp_weather_data[np.isnan(temp_weather_data['Humidity9am'])][['Date','Location']]

### Filling missing data

In [21]:
#### from pandas.tseries.offsets import DateOffset
humidity_mean=temp_weather_data.describe().loc['mean','Humidity9am']
for index, row in missing_humidity9am_data.iterrows():
    date,location = row
    #if the year is the start year, we look up humidity data in the next years 2008, 2009,... (until we find not-null value)
    if date.strftime('%Y')==pd.to_datetime(start_date.loc[start_date['Location']==location,'Date'].values[0]).strftime('%Y'):
        date_offset=date + DateOffset(years=1)
        find_notnull=False
        while date_offset<=pd.to_datetime(end_date.loc[end_date['Location']==location,'Date'].values[0]):
            nextyear_humidity9am = temp_weather_data.loc[(temp_weather_data['Date']==date_offset) & (temp_weather_data['Location']== location),'Humidity9am'].values[0]
            if np.isnan(nextyear_humidity9am)==False :
                temp_weather_data.at[index,'Humidity9am'] = nextyear_humidity9am
                find_notnull=True
                break
            date_offset=date_offset + DateOffset(years=1)
        #if we cannot find not null value
        if find_notnull==False:
            temp_weather_data.at[index,'Humidity9am'] = humidity_mean
            
    #if the year is upper limit year, we look up humidity data in the previous years 2016, 2015,...(until we find not-null value)
    elif date.strftime('%Y')==pd.to_datetime(end_date.loc[end_date['Location']==location,'Date'].values[0]).strftime('%Y'):
        date_offset=date - DateOffset(years=1)
        find_notnull=False
        while date_offset>=pd.to_datetime(start_date.loc[start_date['Location']==location,'Date'].values[0]):
            lastyear_humidity9am =  temp_weather_data.loc[(temp_weather_data['Date']==date_offset) & (temp_weather_data['Location']== location),'Humidity9am'].values[0]
            if np.isnan(lastyear_humidity9am)==False:
                temp_weather_data.at[index,'Humidity9am'] = lastyear_humidity9am
                find_notnull=True
                break
            date_offset=date_offset - DateOffset(years=1)
        #if we cannot find not null value
        if find_notnull==False:
            temp_weather_data.at[index,'Humidity9am'] = humidity_mean

    else:
        filling_missing=0
        date_offset=date + DateOffset(years=1)
        nextyear_humidity9am=-1
        while date_offset<=pd.to_datetime(end_date.loc[end_date['Location']==location,'Date'].values[0]):
            nextyear_humidity9am = temp_weather_data.loc[(temp_weather_data['Date']==date_offset) & (temp_weather_data['Location']== location),'Humidity9am'].values[0]
    
            if np.isnan(nextyear_humidity9am)==False:
                break
            date_offset=date_offset + DateOffset(years=1)
        date_offset= date-DateOffset(years=1)
        lastyear_humidity9am=-1
        while date_offset>=pd.to_datetime(start_date.loc[start_date['Location']==location,'Date'].values[0]):
            lastyear_humidity9am =  temp_weather_data.loc[(temp_weather_data['Date']==date_offset) & (temp_weather_data['Location']== location),'Humidity9am'].values[0]
            if np.isnan(lastyear_humidity9am)==False:
                break
            date_offset=date_offset - DateOffset(years=1)
            
        if np.isnan(nextyear_humidity9am) == True and np.isnan(lastyear_humidity9am)==True:
            filling_missing = humidity_mean
        elif np.isnan(nextyear_humidity9am) == True and np.isnan(lastyear_humidity9am)==False:
            filling_missing = lastyear_humidity9am
        elif np.isnan(nextyear_humidity9am) ==False and np.isnan(lastyear_humidity9am)==True:
            filling_missing = nextyear_humidity9am
        else:
            filling_missing = (nextyear_humidity9am + lastyear_humidity9am)/2
        temp_weather_data.loc[(temp_weather_data['Date']==date) & (temp_weather_data['Location']== location),'Humidity9am'] = filling_missing
        
    

## After preprocessing

### Checking how many null value of Humidity9am after preprocesing


In [22]:
print("Number of humidity 9am data rows that are NaN is ",len(temp_weather_data)-temp_weather_data.describe().loc['count'].loc['Humidity9am'])

Number of humidity 9am data rows that are NaN is  0.0


### Write to processed file

In [32]:
temp_weather_data[['Date','Location','Humidity9am']].to_csv('./data/processed_weather_aus.csv')